README.md
# Kryptiles

[![CircleCI](https://circleci.com/gh/Schultzer/kryptiles.svg?style=svg)](https://circleci.com/gh/Schultzer/kryptiles)


## Examples

```elixir
iex> Kryptiles.random_string(10)
"do77RukqJobZPG3rSJSdCm9JDnX5IT1q"

iex> Kryptiles.random_digits(10)
"3149464061"

iex> Kryptiles.random_bits(10)
<<235, 191>>

iex> Kryptiles.fixed_time_comparison(<<>>, "b0i9XAiBxP")
false

iex> keylen = 20
iex> Kryptiles.pbkdf2("password", "salt", keylen)
<<12, 96, 200, 15, 150, 31, 14, 113, 243, 169, 181, 36, 175, 96, 18, 6, 47, 224,
  55, 166>>
```

## Documentation

[hex documentation for kryptiles](https://hexdocs.pm/kryptiles)


## Installation

```elixir
def deps do
  [{:kryptiles, "~> 0.1.0"}]
end
```

## Acknowledgement

This library was made thanks to [Cryptiles](https://github.com/hapijs/cryptiles) as an implementation in Elixir.

## LICENSE

(The MIT License)

Copyright (c) 2017 Benjamin Schultzer

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## Kryptiles v0.1.0 July 20, 2018

### Enhancements

* Initial


test/kryptiles_test.exs
defmodule KryptilesTest do
  use ExUnit.Case
  use Bitwise

  describe "random_string/1" do
    test "generates the right length string" do
      for n <- 1..1000, do: assert n === byte_size(Kryptiles.random_string(n))
    end

    test "returns an error on invalid bits size" do
      assert Kryptiles.random_string(99999999999999999999) ==  {:error, "failed generating random bits"}
    end
  end

  describe "random_digits/1" do
    test "generates the right length string" do
      for n <- 1..1000, do: assert n === byte_size(Kryptiles.random_digits(n))
    end

    test "returns an error on invalid bits size" do
      assert Kryptiles.random_digits(99999999999999999999) ==  {:error, "failed generating random bits"}
    end

    test "generates equal digits distribution" do
      for d <- digits(), do: assert d in 99000..101000
    end
  end

  def digits(digits \\ [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], size \\ 0)
  def digits(digits, size) when size < 1000000 do
    pos = :erlang.binary_to_integer(Kryptiles.random_digits(1))

    digits
    |> List.replace_at(pos, Enum.at(digits, pos) + 1)
    |> digits(size + 1)
  end
  def digits(digits, _size), do: digits

  describe "random_bits/1" do
    test "returns an error on invalid input" do
      assert Kryptiles.random_bits(0) == {:error, "invalid random bits count"}
    end
  end

  describe "fixed_time_comparison/2" do
    # https://github.com/elixir-lang/plug/blob/master/test/plug/crypto_test.exs#L11-L18
    test "compares binaries securely" do
      assert Kryptiles.fixed_time_comparison(<<>>, <<>>)
      assert Kryptiles.fixed_time_comparison(<<0>>, <<0>>)

      refute Kryptiles.fixed_time_comparison(<<>>, <<1>>)
      refute Kryptiles.fixed_time_comparison(<<1>>, <<>>)
      refute Kryptiles.fixed_time_comparison(<<0>>, <<1>>)
    end
  end

  describe "pbkdf2/4" do
    # https://www.ietf.org/rfc/rfc6070.txt
    test "PBKDF2 with RFC 6070 test vectors" do
      key = Kryptiles.pbkdf2("password", "salt", 20)
      assert Base.encode16(key, case: :lower) === "0c60c80f961f0e71f3a9b524af6012062fe037a6"

      key = Kryptiles.pbkdf2("password", "salt", 20, 2)
      assert Base.encode16(key, case: :lower) === "ea6c014dc72d6f8ccd1ed92ace1d41f0d8de8957"

      key = Kryptiles.pbkdf2("password", "salt", 20, 4096)
      assert Base.encode16(key, case: :lower) === "4b007901b765489abead49d926f721d065a429c1"

      key = Kryptiles.pbkdf2("password", "salt", 20, 16777216)
      assert Base.encode16(key, case: :lower) === "eefe3d61cd4da4e4e9945b3d6ba2158c2634e984"

      key = Kryptiles.pbkdf2("passwordPASSWORDpassword", "saltSALTsaltSALTsaltSALTsaltSALTsalt", 25, 4096)
      assert Base.encode16(key, case: :lower) === "3d2eec4fe41c849b80c8d83662c0e44a8b291a964cf2f07038"

      key = Kryptiles.pbkdf2("pass\0word", "sa\0lt", 16, 4096)
      assert Base.encode16(key, case: :lower) === "56fa6aa75548099dcc37d7f03425e0c3"
    end

    test "PBKDF2 with SHA256" do
      key = Kryptiles.pbkdf2("password", "salt", 32, 32, :sha256)
      assert Base.encode16(key, case: :lower) === "64c486c55d30d4c5a079b8823b7d7cb37ff0556f537da8410233bcec330ed956"
    end

    test "Keylen to long" do
      assert Kryptiles.pbkdf2("password", "salt", bsl(1, 32) + 1) == {:error, "keylen is 4294967297 the maximum keylen is 4294967295"}
    end
  end
end


mix.exs
defmodule Kryptiles.Mixfile do
  use Mix.Project

  @version "0.1.0"

  def project do
    [
      app: :kryptiles,
      version: @version,
      elixir: "~> 1.5",
      name: "Kryptiles",
      source_url: "https://github.com/schultzer/kryptiles",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      description: description(),
      package: package(),
      docs: docs()
    ]
  end

  def application do
    [extra_applications: [:crypto, :logger]]
  end

  defp deps do
    [
      {:ex_doc, "~> 0.14", only: :dev},
    ]
  end

  defp description do
    """
    Elixir implementation of https://github.com/hapijs/cryptiles
    """
  end

  defp package do
    [
      name: :kryptiles,
      maintainers: ["Benjamin Schultzer"],
      licenses: ~w(MIT),
      links: links(),
      files: ~w(CHANGELOG* README* config lib mix.exs)
    ]
end

  def docs do
    [
      source_ref: "v#{@version}",
      main: "readme",
      extras: ["README.md", "CHANGELOG.md"]
    ]
  end

  def links do
    %{
      "GitHub"    => "https://github.com/schultzer/kryptiles",
      "Readme"    => "https://github.com/schultzer/kryptiles/blob/v#{@version}/README.md",
      "Changelog" => "https://github.com/schultzer/kryptiles/blob/v#{@version}/CHANGELOG.md"
    }
  end

end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :ex_cryptiles, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:ex_cryptiles, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/kryptiles.ex
defmodule Kryptiles do
  @moduledoc """
  """

  use Bitwise

  @doc """
  Returns a cryptographically strong pseudo-random data string.
  Takes a size argument for the length of the string.

  ## Examples

      iex> Kryptiles.random_string(0)
      ""

      iex> Kryptiles.random_string(10)
      "do77RukqJobZPG3rSJSdCm9JDnX5IT1q"
  """
  @spec random_string(integer()) :: binary() | {:error, binary()}
  def random_string(size) when is_integer(size) do
    size
    |> Kernel.+(1)
    |> Kernel.*(6)
    |> random_bits()
    |> case do
        {:error, reason} -> {:error, reason}

        bytes ->
          bytes
          |> Base.url_encode64()
          |> String.slice(0, size)
       end
  end

  @doc """
  Returns a cryptographically strong pseudo-random data string consisting of only numerical digits (0-9).
  Takes a size argument for the length of the string.

  ## Examples

      iex> Kryptiles.random_digits(1)
      "9"

      iex> Kryptiles.random_digits(10)
      "3149464061"
  """
  @spec random_digits(integer()) :: binary() | {:error, binary()}
  def random_digits(size) do
    size
    |> Kernel.*(2)
    |> random()
    |> digits(size)
  end

  defp digits(buffer, size, digits \\ [], pos \\ 0)
  defp digits({:error, reason}, _, _, _), do: {:error, reason}
  defp digits(_buffer, size, digits, _pos) when length(digits) == size do
    digits
    |> Enum.reverse()
    |> Enum.join()
  end
  defp digits(buffer, size, digits, pos) when length(digits) < size and pos >= byte_size(buffer) do
    size * 2
    |> random()
    |> digits(size, digits)
  end
  defp digits(buffer, size, digits, pos) when length(digits) < size do
    case :erlang.binary_part(buffer, pos, 1) do
      <<part::integer()>> when part < 250 -> digits(buffer, size, ["#{Integer.mod(part, 10)}" | digits], pos + 1)

      _                                   -> digits(buffer, size, digits, pos + 1)
    end
  end

  @doc """
  Returns a cryptographically strong pseudo-random bytes
  Takes a size argument for the length of the string.

  ## Examples

      iex> Kryptiles.random_bits(1)
      <<236>>

      iex> Kryptiles.random_bits(10)
      <<235, 191>>
  """
  @spec random_bits(integer()) :: binary() | {:error, binary()}
  def random_bits(bits) when bits <= 0, do: {:error, "invalid random bits count"}
  def random_bits(bits) do
    bits
    |> Kernel./(8)
    |> Float.ceil()
    |> Kernel.round()
    |> random()
  end

  @doc false
  @spec random(integer()) :: binary() | {:error, binary()}
  def random(bytes) do
    try do
      :crypto.strong_rand_bytes(bytes)
    rescue
      ArgumentError -> {:error, "failed generating random bits"}
    end
  end

  # https://github.com/elixir-lang/plug/blob/master/lib/plug/crypto.ex#L94-L114
  # http://codahale.com/a-lesson-in-timing-attacks/
  @doc """
  Compare two strings using fixed time algorithm (to prevent time-based analysis of MAC digest match).
  Returns `true` if the strings match, `false` if they differ.

  ## Examples

      iex> Kryptiles.fixed_time_comparison(<<>>, <<>>)
      true

      iex> Kryptiles.fixed_time_comparison(<<>>, "b0i9XAiBxP")
      false
  """
  @spec fixed_time_comparison(binary(), binary()) :: true | false
  def fixed_time_comparison(left, right) when byte_size(left) == byte_size(right) do
    __fixed_time_comparison__(left, right) == 0
  end
  def fixed_time_comparison(_left, _right), do: false

  defp __fixed_time_comparison__(left, right, acc \\ 0)
  defp __fixed_time_comparison__(<<>>, <<>>, acc), do: acc
  defp __fixed_time_comparison__(<<x, left::binary()>>, <<y, right::binary()>>, acc) do
    __fixed_time_comparison__(left, right, acc ||| (x ^^^ y))
  end

  # bsl(1, 32) - 1 === 4294967295
  @doc """
  Computes a pbkdf2 bitstring [RFC 2898](https://tools.ietf.org/html/rfc2898)

  * `options` are an `Enumerable.t()` with these keys:
    * `digest` is any of `:md5 | :sha | :sha224 | :sha256 | :sha384 | :sha512 ` defaults to `:sha`
    * `iterations` is a `non_neg_integer()` defaults to `1`


  ## Examples

      iex> keylen = 20
      iex> Kryptiles.pbkdf2("password", "salt", keylen)
      <<12, 96, 200, 15, 150, 31, 14, 113, 243, 169, 181, 36, 175, 96, 18, 6, 47, 224,
        55, 166>>
  """
  @spec pbkdf2(binary(), binary(), pos_integer(), pos_integer, atom()) :: binary() | {:error, binary()}
  def pbkdf2(password, salt, keylen, iterations \\ 1, digest \\ :sha)
  def pbkdf2(_password, _salt, keylen, _iterations, _digest) when not is_integer(keylen) do
    {:error, "invalid keylen: #{inspect keylen}"}
  end
  def pbkdf2(_password, _salt, keylen, _iterations, _digest) when keylen > 4294967295 do
    {:error, "keylen is #{inspect keylen} the maximum keylen is 4294967295"}
  end
  def pbkdf2(_password, _salt, _keylen, iterations, _digest) when not is_integer(iterations) do
    {:error, "invalid iterations: #{inspect iterations}"}
  end
  for digest <- ~w(md5 sha sha224 sha256 sha384 sha512)a do
    def pbkdf2(password, salt, keylen, iterations, unquote(digest)), do: __pbkdf2__(mac_fun(unquote(digest), password), salt, keylen, iterations)
  end
  def pbkdf2(_password, _salt, _keylen, _iterations, digest), do: {:error, "unknown digest: #{inspect digest}"}

  defp __pbkdf2__(fun, salt, keylen, iterations, block_index \\ 1, length \\ 0, acc \\ [])
  defp __pbkdf2__(_fun, _salt, keylen, _iterations, _block_index, length, acc) when length >= keylen do
    acc
    |> :erlang.iolist_to_binary()
    |> :erlang.binary_part(0, keylen)
  end
  defp __pbkdf2__(fun, salt, keylen, iterations, block_index, length, acc) do
    initial = fun.(<<salt::binary, block_index::integer-size(32)>>)
    block = iterate(fun, iterations - 1, initial, initial)
    __pbkdf2__(fun, salt, keylen, iterations, block_index + 1, byte_size(block) + length, [acc | block])
  end

  defp iterate(_fun, 0, _prev, acc), do: acc
  defp iterate(fun, iteration, prev, acc) do
    next = fun.(prev)
    iterate(fun, iteration - 1, next, :crypto.exor(next, acc))
  end

  defp mac_fun(digest, secret) do
    &:crypto.hmac(digest, secret, &1)
  end
end
